src/OpenSage.Game/Graphics/ParticleSystems/ParticleSystem.cs

```diff
@@ -432,6 +403,26 @@ private void UpdateParticle(ref Particle particle)
             particle.Timer += 1;
         }
 
+        private static void FindKeyframes<T>(int timer,
+            IReadOnlyList<T> keyFrames,
+            out T next, out T prev)
+            where T : struct, IParticleKeyframe
```

Comment:
This function is generic to avoid boxing the results.

Reply:
That's interesting - are you sure it doesn't box when it calls the `Time` property? Since it knows about that property through an interface, I wasn't sure if it would box the struct, even though because of the generic type it could access it directly.